File: src/clut_header.rs
use crate::binary::NetString;
use crate::types::{ClutVersion, CompressType, GameVersion, PatchVersion, PlatformId};
use binrw::{BinRead, BinWrite};

/// CLUT file header containing metadata and compression information
#[derive(Debug, Clone, BinRead, BinWrite)]
#[br(little)]
#[bw(little)]
pub struct ClutHeader {
    /// Magic bytes (0xDF23)
    #[br(assert(magic == 0xDF23, "Invalid magic bytes: expected 0xDF23, got {:#X}", magic))]
    pub magic: u16,

    /// File version
    #[br(assert(file_version == ClutVersion::SeparateVersioning, "Unsupported version: {:?}", file_version))]
    pub file_version: ClutVersion,

    /// Compression type used for the data
    pub compression: CompressType,

    /// Target platform
    pub platform: PlatformId,

    /// Repository name
    pub repository: NetString,

    /// Game version
    pub version: NetString,

    /// Patch version
    pub patch_version: NetString,

    /// Base patch URL (can be empty)
    pub base_patch_url: NetString,

    /// Size of decompressed data
    pub decompressed_size: i32,

    /// Size of compressed data (only present if compression != None)
    #[br(if(compression != CompressType::None))]
    #[bw(if(*compression != CompressType::None))]
    pub compressed_size: Option<i32>,
}

impl ClutHeader {
    /// Parse the version strings into structured types
    pub fn parse_versions(&self) -> anyhow::Result<(GameVersion, PatchVersion)> {
        let game_version = GameVersion::new(&self.version.0)?;
        let patch_version = PatchVersion::new(&self.patch_version.0)?;
        Ok((game_version, patch_version))
    }

    /// Get the actual compressed size (falls back to decompressed size if no compression)
    pub fn get_compressed_size(&self) -> i32 {
        self.compressed_size.unwrap_or(self.decompressed_size)
    }

    /// Check if the base patch URL is meaningful (not empty or whitespace)
    pub fn has_base_patch_url(&self) -> bool {
        !self.base_patch_url.0.trim().is_empty()
    }
}

impl Default for ClutHeader {
    fn default() -> Self {
        Self {
            magic: 0xDF23,
            file_version: ClutVersion::SeparateVersioning,
            compression: CompressType::None,
            platform: PlatformId::Win32,
            repository: NetString("UNKNOWN".to_string()),
            version: NetString(GameVersion::epoch().to_string()),
            patch_version: NetString(PatchVersion::epoch().to_string()),
            base_patch_url: NetString(String::new()),
            decompressed_size: 0,
            compressed_size: None,
        }
    }
}

File: dotnet/FFXIVDownloader/Lut/ClutHeader.cs
using FFXIVDownloader.Thaliak;
using System.Diagnostics.CodeAnalysis;
using static FFXIVDownloader.ZiPatch.Config.ZiPatchConfig;

namespace FFXIVDownloader.Lut;

public sealed class ClutHeader
{
    public const ushort MAGIC = 0xDF23;

    public ClutVersion FileVersion { get; }
    public CompressType Compression { get; set; }
    public PlatformId Platform { get; init; }

    public required string Repository { get; set; }
    public required GameVersion Version { get; set; }
    public required PatchVersion PatchVersion { get; set; }
    public required string? BasePatchUrl { get; set; }

    public int DecompressedSize { get; set; }
    public int CompressedSize { get; set; }

    [SetsRequiredMembers]
    public ClutHeader()
    {
        FileVersion = ClutVersion.SeparateVersioning;
        Platform = PlatformId.Win32;
        Repository = "UNKNOWN";
        Version = GameVersion.Epoch;
        PatchVersion = PatchVersion.Epoch;
        BasePatchUrl = null;
    }

    [SetsRequiredMembers]
    public ClutHeader(BinaryReader reader)
    {
        var magic = reader.ReadUInt16();
        if (magic != MAGIC)
            throw new LutException($"Invalid magic: {magic:X4}");

        FileVersion = (ClutVersion)reader.ReadUInt16();
        if (FileVersion != ClutVersion.SeparateVersioning)
            throw new LutException($"Unsupported version: {FileVersion}");

        Compression = (CompressType)reader.ReadByte();
        Platform = (PlatformId)reader.ReadByte();

        Repository = reader.ReadString();
        Version = new GameVersion(reader.ReadString());
        PatchVersion = new PatchVersion(reader.ReadString());
        BasePatchUrl = reader.ReadString();
        if (string.IsNullOrWhiteSpace(BasePatchUrl))
            BasePatchUrl = null;

        DecompressedSize = reader.ReadInt32();
        if (Compression != CompressType.None)
            CompressedSize = reader.ReadInt32();
    }

    public void Write(BinaryWriter writer)
    {
        writer.Write(MAGIC);
        writer.Write((ushort)FileVersion);
        writer.Write((byte)Compression);
        writer.Write((byte)Platform);
        writer.Write(Repository);
        writer.Write(Version.ToString());
        writer.Write(PatchVersion.ToString());
        writer.Write(BasePatchUrl ?? string.Empty);
        writer.Write(DecompressedSize);
        if (Compression != CompressType.None)
            writer.Write(CompressedSize);
    }
}
